src/vs/workbench/contrib/chat/common/chatWordCounter.ts

```diff
@@ -10,13 +10,43 @@ export interface IWordCountResult {
 	isFullString: boolean;
 }
 
+const r = String.raw;
+
+/**
+ * Matches `[text](link title?)` or `[text](<link> title?)`
+ *
+ * Taken from vscode-markdown-languageservice
+ */
+const linkPattern =
+	r`(?<!\\)` + // Must not start with escape
+
+	// text
+	r`(!?\[` + // open prefix match -->
+	/**/r`(?:` +
+	/*****/r`[^\[\]\\]|` + // Non-bracket chars, or...
+	/*****/r`\\.|` + // Escaped char, or...
+	/*****/r`\[[^\[\]]*\]` + // Matched bracket pair
+	/**/r`)*` +
+	r`\])` + // <-- close prefix match
+
+	// Destination
+	r`(\(\s*)` + // Pre href
+	/**/r`(` +
+	/*****/r`[^\s\(\)<](?:[^\s\(\)]|\([^\s\(\)]*?\))*|` + // Link without whitespace, or...
+	/*****/r`<(?:\\[<>]|[^<>])+>` + // In angle brackets
+	/**/r`)` +
+
+	// Title
+	/**/r`\s*(?:"[^"]*"|'[^']*'|\([^\(\)]*\))?\s*` +
+	r`\)`;
+
 export function getNWords(str: string, numWordsToCount: number): IWordCountResult {
 	// This regex matches each word and skips over whitespace and separators. A word is:
 	// A markdown link
 	// One chinese character
 	// One or more + - =, handled so that code like "a=1+2-3" is broken up better
 	// One or more characters that aren't whitepace or any of the above
-	const allWordMatches = Array.from(str.matchAll(/\[([^\]]+)\]\(([^)]+)\)|\p{sc=Han}|=+|\++|-+|[^\s\|\p{sc=Han}|=|\+|\-]+/gu));
+	const allWordMatches = Array.from(str.matchAll(new RegExp(linkPattern + r`|\p{sc=Han}|=+|\++|-+|[^\s\|\p{sc=Han}|=|\+|\-]+`, 'gu')));
```

Comment:
Thanks, can you add a test case for whatever case you fixed?

Reply:
Good idea! Just added some tests for the parens and brackets issues